File: JAVA/203 - Remove Linked List Elements.java
/*
 203. Remove Linked List Elements (E)

 Approach 1 : One pointer
 Check if node's next exist, if exist, compare two nodes val and relink. 
 Time: O(n)
 Space: O(1)

*/

/* Approach 1 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode node = dummy;
        while (node.next != null) {
            if (node.next.val == val)
                node.next = node.next.next;
            else node = node.next;
        }
        return dummy.next;
    }
}

File: C++/203-Remove_Linked_List_Elements.cpp
/*
 203. Remove Linked List Elements (Easy)

 Notes:
    One pointers & Prenode - O(n)
    Use a pointer to scan the list. If the node is duplicate, delete it and connect its previous node to the next one. 
*/

class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        head = dummy;
        while (head->next)
        {
            if (head->next->val == val)
            {
                ListNode* del = head->next;
                head->next = del->next;
                delete del;
            }
            else
                head = head->next;
        }
        return dummy->next;
    }
};
